feat : FeelinApp에 맞는 커스텀 AlertDialog기능 & UI 구현

inwoodev opened the pull request:
## PR 타입
어떤 변경에 대한 PR인가요?

<!-- 아래 체크리스트 중 해당 되는 부분에 체크해주세요 "x". -->

- [ ] 버그 수정
- [x] Feature / 신기능
- [ ] 코드 스타일 업데이트 (formatting, local variables)
- [ ] 리팩토링 (no functional changes, no api changes)
- [ ] 빌드 관련 변경사항
- [ ] CI 관련된 변경사항
- [ ] Documentation content changes
- [ ] 구조 변경
- [ ] Other... 추가 설명 요망:


## 배경
로그인 실패, 네트워크 오류 등 상황에서 유저에게 이를 알릴 수 있는 Feelin의 커스텀 다이얼로그가 필요합니다.

## 목표
- 커스텀 alertview 구현
- Figma와 동일한 UI 구현
- 다크모드 대응

## 결과 (Optional)
close #37

- Alert with two buttons
![Simulator Screen Recording - iPhone 15 Pro - 2024-07-11 at 00 00 24](https://github.com/project-lyrics/app-iOS/assets/69072471/857cfa62-a315-4e62-b674-cc05d178a92c)

- Alert with a single button
![Simulator Screen Recording - iPhone 15 Pro - 2024-07-11 at 00 02 57](https://github.com/project-lyrics/app-iOS/assets/69072471/7edd3013-84f2-46e2-8ce4-8e8a068915f0)

## Trouble or Trouble Shooting(Optional)

<!-- 해당 PR에서 발생한 문제를 해결한 내용을 작성해주세요.  !-->
- button의 백그라운드 색상을 변경하기 위해 `UIColor`를 `UIImage`로 변환하는 기능을 추가하였습니다.

해당 AlertDialog까 present한 상황에서 darkmode를 변경할 경우 버튼 배경색상이 실시간으로 변경되지 않는 문제가 있었습니다.

해당 문제를 해결하기 위해서 부득이하게 `FeelinAlertViewController`의 `willTransition` 메서드에서 trait collection이 변경 될 때 마다 배경화면을 재적용하는 메서드를 다음과 같이 추가하였습니다:

``` swift

public override func willTransition(to newCollection: UITraitCollection, with coordinator: any UIViewControllerTransitionCoordinator) {
        super.willTransition(to: newCollection, with: coordinator)
        
        // MARK: - 버튼에 배경색에 다크모드를 실시간 적용하기 위해 setBackgroundImage를 해당 메서드에서 호출
        
        self.leftButton.setBackgroundImage(Colors.modal.image(), for: .normal)
        self.rightButton.setBackgroundImage(Colors.modal.image(), for: .normal)
        self.singleButton.setBackgroundImage(Colors.modal.image(), for: .normal)
        self.singleButton.setBackgroundImage(Colors.gray01.image(), for: .highlighted)
        self.leftButton.setBackgroundImage(Colors.gray01.image(), for: .highlighted)
        self.rightButton.setBackgroundImage(Colors.gray01.image(), for: .highlighted)
        
        
        // MARK: - 디자이너 요청에 따라 cancelButton만 다크모드시 gray06색상 적용
        
        switch newCollection.userInterfaceStyle {
        case .light:
            self.leftButton.setTitleColor(Colors.gray02, for: .normal)
            
        case .dark:
            self.leftButton.setTitleColor(Colors.gray06, for: .normal)
        default:
            return
        }
    }
```

inwoodev committed:
move : BottomSheetViewController를 ViewControllers group으로 이동

inwoodev committed:
refactor : UIControl+ -> UIControl+Combine으로 이름 수정

- Combine 프레임워크에 특화된 extension이기에 네이밍 수정

inwoodev committed:
feat : UIColor를 image로 변환할 수 있는 기능 추가

inwoodev committed:
feat : FeelinAlertViewController 컴포넌트 구현

- 다크모드 대응할 수 있게 UI구현
- Single Button, two button에 대응하기 용이하게 설계
- FeelinAlertViewController를 사용하기 용이가헤 할 수 있는 UIViewController extension 추가

inwoodev commented on Projects/Shared/DesignSystem/Sources/Extensions/UIViewController+.swift:
```diff
@@ -0,0 +1,102 @@
+            )
+        }
+    
+    public func showAlert(
+        title: String,
+        message: String,
+        attributedMessage: NSAttributedString? = nil,
+        singleActionTitle: String,
+        actionCompletion: (() -> Void)? = nil
+    ) {
+        let alertViewController = FeelinAlertViewController(
+            titleText: title,
+            messageText: message,
+            attributedMessageText: attributedMessage
+        )
+        
+        showAlert(
+            alertViewController: alertViewController,
+            actionTitle: singleActionTitle,
+            actionCompletion: actionCompletion
+        )
+    }
```
alert dialog에서 버튼이 하나만 필요할 경우 해당 메서드를 호출할 수 있습니다. 다음은 예시 코드입니다:

``` swift
output.loginResult
  .receive(on: DispatchQueue.main)
  .sink { [weak self] result in
     switch result {
        case .success:
          self?.coordinator?.didFinish()
        case .failure(let error):
          self?.showAlert(title: "로그인 에러", message: "로그인 과정에서 문제가 생겼어요", singleActionTitle: "확인", actionCompletion: {
          print("action completion")
        })
      }
    }
    .store(in: &cancellables)
```

inwoodev commented on Projects/Shared/DesignSystem/Sources/Extensions/UIViewController+.swift:
```diff
@@ -0,0 +1,102 @@
+        )
+    }
+    
+    public func showAlert(
+        contentView: UIView,
+        leftActionTitle: String = "취소",
+        rightActionTitle: String = "확인",
+        leftActionCompletion: (() -> Void)? = nil,
+        rightActionCompletion: (() -> Void)? = nil
+    ) {
+        let alertViewController = FeelinAlertViewController(contentView: contentView)
+        
+        showAlert(
+            alertViewController: alertViewController,
+            leftActionTitle: leftActionTitle,
+            rightActionTitle: rightActionTitle,
+            leftActionCompletion: leftActionCompletion,
+            rightActionCompletion: rightActionCompletion
+        )
+    }
```
버튼이 두 개가 필요할 경우 해당 메서드를 호출할 수 있습니다.

inwoodev commented on Projects/Shared/DesignSystem/Sources/ViewControllers/FeelinAlertViewController.swift:
```diff
@@ -0,0 +1,309 @@
+        }
+    }
+    
+    public override func willTransition(to newCollection: UITraitCollection, with coordinator: any UIViewControllerTransitionCoordinator) {
+        super.willTransition(to: newCollection, with: coordinator)
+        
+        // MARK: - 버튼에 배경색에 다크모드를 실시간 적용하기 위해 setBackgroundImage를 해당 메서드에서 호출
+        
+        self.leftButton.setBackgroundImage(Colors.modal.image(), for: .normal)
+        self.rightButton.setBackgroundImage(Colors.modal.image(), for: .normal)
+        self.singleButton.setBackgroundImage(Colors.modal.image(), for: .normal)
+        self.singleButton.setBackgroundImage(Colors.gray01.image(), for: .highlighted)
+        self.leftButton.setBackgroundImage(Colors.gray01.image(), for: .highlighted)
+        self.rightButton.setBackgroundImage(Colors.gray01.image(), for: .highlighted)
+        
+        
+        // MARK: - 디자이너 요청에 따라 cancelButton만 다크모드시 gray06색상 적용
+        
+        switch newCollection.userInterfaceStyle {
+        case .light:
+            self.leftButton.setTitleColor(Colors.gray02, for: .normal)
+            
+        case .dark:
+            self.leftButton.setTitleColor(Colors.gray06, for: .normal)
+        default:
+            return
+        }
+    }
+    
```
TroubleShooting에서 언급한 부분입니다.

derrickkim0109 commented on Projects/Shared/DesignSystem/Sources/ViewControllers/FeelinAlertViewController.swift:
```diff
@@ -0,0 +1,309 @@
+    }()
+    
+    private lazy var buttonStackView: UIStackView = {
+        let view = UIStackView()
+        view.distribution = .fillEqually
```
Spacing을 따로 설정하지 않아도 될까요?

inwoodev replied:
넵! 피그마상 별도의 spacing이 없더라구요!

derrickkim0109 commented on Projects/Shared/DesignSystem/Sources/ViewControllers/FeelinAlertViewController.swift:
```diff
@@ -0,0 +1,309 @@
+        buttonStackView.addArrangedSubview(leftButton)
+        buttonStackView.addArrangedSubview(rightButton)
+        buttonStackView.addArrangedSubview(singleButton)
+        ///
```
이 부분은 제거해도 되지 않을까요?

inwoodev replied:
실수로 주석을 추가 해 버렸네요 ㅠ 꼭 필요한 코드여서 일단 주석만 지우겠습니다~!

derrickkim0109 reviewed (approved):
수고하셨습니다. spacing 값 처리만 확인 되시면 될 것 같아서 승인처리 해두겠습니다~!

inwoodev committed:
remove : 불필요한 주석 제거

##### 제목은 최대 우측 '|' 까지만 입력 ######## -> |

# 본문은 위에 작성
##########본문은 우측 '|' 까지만 입력 ######################## -> |

# 꼬릿말은 아래에 작성: Ex #issue number

# --- COMMIT END ---
